finance-admin-api/api/server.go

```diff
@@ -23,18 +25,23 @@ type Dispatch interface {
 
 type FileStorage interface {
 	GetFile(ctx context.Context, bucketName string, filename string, versionID string) (*s3.GetObjectOutput, error)
-	PutFile(ctx context.Context, bucketName string, fileName string, file io.Reader) error
+	PutFile(ctx context.Context, bucketName string, fileName string, file io.Reader) (*string, error)
 	FileExists(ctx context.Context, bucketName string, filename string, versionID string) bool
 }
 
+type Reports interface {
+	Generate(ctx context.Context, filename string, reportQuery db.ReportQuery) (*os.File, error)
```

Comment:
It's a bit of a nitpick, but the convention in Go is to have the interface declared where it is used rather than with its implementation. So here, instead of doing `db.ReportQuery`, the `ReportQuery` interface should be declared in this file instead. That way, this file doesn't even need to import the db package.

Reply:
I'm not sure this is possible - because ReportQuery is used in the db package, if we were to define it here in api, we'd import db from here, but then db would need to import api to use the definition of ReportQuery, creating an import cycle. I've just tried to do it locally and it's refusing to build

Reply:
I _think_ the idiomatic Go approach would be to have an interface in each package it is used. In OOP languages (e.g. PHP), an interface is a blueprint for a class to adhere to, whereas in Go, it is the contract the consuming function accepts. So each place it is accepted should define its own contract (interface). Don't think there's much benefit in getting bogged down with this here though.